packages/@aws-cdk/aws-servicecatalog/lib/private/association-manager.ts

```diff
@@ -0,0 +1,123 @@
+    InputValidator.validateLength(this.generateAssocationString(scope, pair), 'description', 0, 2000, options.description);
+    if (!this.associationMap.get(associationKey).get(Constraints.RESOURCE_UPDATE)) {
+      const constraint = new CfnResourceUpdateConstraint(scope, `ResourceUpdateConstraint${associationKey}`, {
+        acceptLanguage: options.acceptLanguage,
+        description: options.description,
+        portfolioId: pair.portfolio.portfolioId,
+        productId: pair.product.productId,
+        tagUpdateOnProvisionedProduct: options.tagUpdateOnProvisionedProductAllowed || options.tagUpdateOnProvisionedProductAllowed === undefined
+          ? Allowed.ALLOWED : Allowed.NOT_ALLOWED,
+      });
+
+      constraint.addDependsOn(this.associationMap.get(associationKey).get(Constraints.ASSOCIATION));
+      this.associationMap.get(associationKey).set(Constraints.RESOURCE_UPDATE, constraint);
+    } else {
+      throw new Error(`Cannot have multiple resource update constraints for association ${this.generateAssocationString(scope, pair)}`);
+    }
+  }
+
+  public static associateTagOption(scope: cdk.Resource, resourceId: string, tagOption: TagOption) {
+    Object.keys(tagOption).forEach(key => {
+      InputValidator.validateLength(resourceId, 'TagOption key', 1, 128, key);
+      tagOption[key].forEach(value => {
+        InputValidator.validateLength(resourceId, 'TagOption value', 1, 256, value.value);
+        const tagOptionKey = hashValues(key, value.value);
+        if (!this.tagOptionMap.has(tagOptionKey)) {
+          const tO = new CfnTagOption(scope, `TagOption${tagOptionKey}`, {
+            key: key,
+            value: value.value,
+            active: value.active ?? true,
+          });
+          this.tagOptionMap.set(tagOptionKey, tO);
+        }
+
+        new CfnTagOptionAssociation(scope, `TagOptionAssociation${hashValues(scope.node.addr, tagOptionKey)}`, {
+          resourceId: resourceId,
+          tagOptionId: this.tagOptionMap.get(tagOptionKey)!.ref,
+        });
+      });
+    });
+  }
+
+  private static associationMap = new Map<string, any>();
+  private static tagOptionMap = new Map<string, CfnTagOption>();
+
+  private static associationPrecheck(scope: cdk.Resource, product: IProduct): [AssociationPair, string] {
+    const pair = this.resolveProductAndPortfolio(scope, product);
+    const associationKey = hashValues(pair.portfolio.node.addr, pair.product.node.addr, scope.stack.node.addr);
+    if (!this.associationMap.has(associationKey)) {
+      this.associateProductWithPortfolio(scope, pair.portfolio, pair.product);
+    }
+    return [pair, associationKey];
+  }
+
+  private static resolveProductAndPortfolio(scope: cdk.Resource, product: IProduct): AssociationPair {
+    return {
+      portfolio: scope as Portfolio,
+      product: product as Product,
+    };
+  }
+
+  private static generateAssocationString(scope: cdk.Resource, pair: AssociationPair) {
```

Comment:
This is a really weird method... Is this some way to describe the resources whose validation is failing? Because if it, we need to rename this method.

Reply:
we use a notation of `portfolio | product` for describing associations, and wanted to maintain that in our error messages. It isn't specifically for throwing errors, it just generates a nicely formatted string for it and we call it in the errors to avoid duplicate code.

Reply:
maybe `prettyPrintAssociation` ?

Reply:
I like it.